Extend spacemacs/copy-file command with extra actions

dalanicolai opened the pull request:
This PR implements an alternative for PR #8974.

lebensterben commented on layers/+spacemacs/spacemacs-defaults/funcs.el:
```diff
@@ -1350,10 +1350,43 @@ the right."
+               (not (y-or-n-p (format "File %s exists. Overwrite?: " new-file-name))))
+      (error "Canceled"))
+    (pcase
+        (read-answer "New file action: "
```
In general avoid string comparison. This is inefficient.

You can just match the character code.

If you really want, match symbol, e.g. 'only-write.

dalanicolai replied:
Thanks for this info, handy to be aware of. But in this case the code only runs once only when this command is called, so I guess the efficiency is not so important in this case.

So otherwise you would suggest to compare only the first (or other relevant) characters or otherwise compare with `intern`?

dalanicolai replied:
Ah yes, I did not know about `user-error`. Indeed I used a catch-throw before, but I thought `error` looks nicer (and then accept the "wrong" behavior). Very useful info!

lebensterben replied:
> So otherwise you would suggest to compare only the first (or other relevant) characters, or alternatively compare by converting to symbol with `intern`?

See https://www.gnu.org/software/emacs/manual/html_node/elisp/Multiple-Queries.html

- wrap this in a let form to set read-answer-short to t. 
- Change the long answer to a single character string. the long answer is not shown but is returned. Since it has length 1, it's fast to compare.
- Since read-answer-short is t, user only need to type one key to select a choice, instead of being required to type a sentence/phrase

dalanicolai replied:
Yes I was aware of this, but then I would probably have wrapped it in `read-answer-short 'auto` (and indeed make the long answers slightly shorter, but still a little more informative)

lebensterben commented on layers/+spacemacs/spacemacs-defaults/funcs.el:
```diff
@@ -1350,10 +1350,43 @@ the right."
+    (when (directory-name-p new-file-name)
+      (if buffer-file-name
+          (setq new-file-name (concat new-file-name (buffer-name)))
+        (error "Copying of non-visiting file buffers requires to specify new file name.")))
```
"non-visiting file" is not easy to understand

dalanicolai replied:
I agree somewhat, although I think it is hard to misunderstand this (of course it is "non-visiting file buffer").
I would be happy if you could suggest a better term (English is not my native language).

dalanicolai replied:
Maybe "non-visiting-file buffers"?

lebensterben replied:
this feels like romance language, not english

lebensterben commented on layers/+spacemacs/spacemacs-defaults/funcs.el:
```diff
@@ -1350,10 +1350,43 @@ the right."
+                       ("quit" ?q "exit")))
+      ("quit" (print "Canceled"))
+      (code (let ((dir (file-name-directory new-file-name)))
+              (when (not (file-exists-p dir))
```
unless

dalanicolai replied:
Good point. Interesting how I miss these things regularly...

lebensterben commented on layers/+spacemacs/spacemacs-defaults/funcs.el:
```diff
@@ -1350,10 +1350,43 @@ the right."
+                (make-directory dir t)))
+            (if (string= "replace" code)
+                (write-file new-file-name))
+            (write-region (point-min) (point-max) new-file-name)
```
write-region after write-file?

not indented correctly

dalanicolai replied:
Good catch! And thanks for your review :+1:

lebensterben reviewed (changes requested):
Also, replace error by user-error. These errors should not start the debugger.

lebensterben commented:
key-bindings for the multiple choice can be improved.

If this command is invoked via SPC f c, we can put the most possible choice under the key c.
So the user just quickly type SPC f c c to perform the default behavior.

I suggest
- c Copy and open in new buffer
- C Copy and open on current buffer
- d Don't open the new file after copying
- q Quit

dalanicolai commented:
I agree with you @lebensterben that generally, it is nice to put the default bindings on the quickest way to type. However, in this case after typing `SPC f c` the user first has to enter a new filename and a RET. So it is not very important here that the default action is under `c`.
Also, I always prefer downcase letters, so I would vote against using `C`.

On the other hand, I do agree that my suggestions (r, n and o) might not be the best choices. But I would suggest we choose wither the most sensible mnemonic or most consistent with other commands keybindings (which should generally most sensible mnemonic also).

So another main candidate for me was:
`c` for open in current buffer
`o` for open in other buffer
 `w` for just write (or 's' for just save)

lebensterben commented:
okay. c o w looks a good choice

dalanicolai commented:
Updated! The reason why I chose "replace" over "current-buffer" is that the former "more explicitly" implies that the old-file buffer is getting killed (or well... replaced). But I guess this `cow` will also produce our milk...

lebensterben commented on layers/+spacemacs/spacemacs-defaults/funcs.el:
```diff
@@ -1350,10 +1350,43 @@ the right."
+This command prompts for a new filename then prompts for
+selecting an action to perform on the new file.
+
+If prefixed with the universal ARG \\[universal-argument] the
```
Can you remind me why we need this prefix argument?
```elisp
(read-file-name "Save file as: " (if arg buffer-file-name default-directory))
```

Can we just write
```elisp
(read-file-name "Save file as: " (or buffer-file-name default-directory))
```

dalanicolai replied:
This argument is to include the filename in the prompt for new filename like in the [https://github.com/syl20bnr/spacemacs/pull/8974#issuecomment-755299650](original proposal) (but then inverted). The `or` wouldn't work here, at least not for how I meant it. On the other hand `(buffer-name)` would be even better than buffer-file-name.

lebensterben commented:
Use this
```elisp
(defun spacemacs/save-as (filename &optional visit)
  "Save current buffer or active region as specified file.
When called interactively, it first prompts for FILENAME, and then asks
 whether to VISIT it, and if so, whether to show it in current window or
 another window.

FILENAME  a non-empty string as the name of the saved file.
VISIT     When it's `:current', open FILENAME in current window. When it's
          `:other', open FILENAME in another window. When it's nil, only
          save to FILENAME but does not visit it. (Default to `:current'
          when called from a LISP program.)

When FILENAME already exists, it also asks the user whether to overwrite it."
  (interactive (let* ((filename (expand-file-name (read-file-name "Save buffer as: ")))
                      (choices  '("Current window"
                                  "Other window"
                                  "Don't open"))
                      (actions  '(:current :other nil))
                      (visit    (let ((completion-ignore-case t))
                                  (nth (cl-position
                                        (completing-read "Do you want to open the file? "
                                                         choices nil t)
                                        choices
                                        :test #'equal)
                                       actions))))
                 (list filename visit)))
  (unless (called-interactively-p 'any)
    (cl-assert (and (stringp filename)
                    (not (string-empty-p filename))
                    (not (directory-name-p filename)))
              t "Expect an non-empty filepath, found: %s")
    (setq filename (expand-file-name filename)
          visit (or visit :other))
    (let ((choices '(:current :other nil)))
      (cl-assert (memq visit choices)
                 t "Found %s, expect one of %s")))
  (let ((dir (file-name-directory filename)))
    (unless (file-directory-p dir)
      (make-directory dir t)))
  (if (use-region-p)
      (write-region (region-beginning) (region-end) filename nil nil nil t)
    (write-region nil nil filename nil nil nil t))
  (pcase visit
    (:current (find-file filename))
    (:other   (funcall-interactively 'find-file-other-window filename))))
```

dalanicolai commented:
Haha! That's funny... it is a full rewrite. Anyway, that is fine with me. However, this doesn't yet include an option to include the (current) filename in the prompt.
Also it is not possible to select a "choice" with a single keystroke, but that is a matter of preference of course (I like this solution too).

lebensterben commented:
@dalanicolai 
That's a bug of helm.
If you want, `(read-file-name "Save buffer as: " nil nil nil buffer-file-name)`

lebensterben commented:
I already found this issue. I just didn't bother to fix it.

lebensterben commented:
it's possible to select with two keystroke. the initial and <enter>.

lebensterben commented:
Also note that I've edited the code snippet. It's certainly not final.

lebensterben reviewed (changes requested):
I'm going to merge this. Just fix the docstring.
Also please update CHANGELOG.develop

dalanicolai commented:
I got a little confused about 5d213ce. It caused this PR to consist of 2 commits, and I think it was not meant like that. Anyway, I reverted to a single commit and rebased the PR on develop.

lebensterben commented:
https://github.com/syl20bnr/spacemacs/pull/14754/files
This is what the changes will be like when it's squashed and merged.

dalanicolai commented:
You mean it is fine to have multiple commits because you will squash them when merging?

lebensterben commented:
yes

dalanicolai commented on layers/+spacemacs/spacemacs-defaults/funcs.el:
```diff
@@ -1350,10 +1350,55 @@ the right."
-  (call-interactively 'write-file))
+(defun spacemacs/save-as (filename &optional visit)
+  "Save current buffer or active region as specified file.
+ When called interactively, it first prompts for FILENAME, and then asks
```
Done

lebensterben commented:
Merged. Thanks for your contributions.

bcc32 commented:
I think this key binding needs to be updated, since the name of the function was changed:

```elisp
;; file -----------------------------------------------------------------------
(spacemacs|spacebind
 "Files manipulation."
 :global
 (("f" "Files"
   ("A" spacemacs/find-file-and-replace-buffer "Set another file for buffer...")
   ("c" spacemacs/copy-file "Copy file to new file...")
```
